features/device-config/service/src/main/java/org/opennms/features/deviceconfig/service/impl/DeviceConfigServiceImpl.java

```diff
@@ -81,50 +97,98 @@ public void triggerConfigBackup(String ipAddress, String location, String config
+    public CompletableFuture<DeviceConfig> getDeviceConfig(String ipAddress, String location, String configType, int timeout) throws IOException {
         return pollDeviceConfig(ipAddress, location, configType)
                 .orTimeout(timeout, TimeUnit.MILLISECONDS)
-                .thenApply(resp -> {
-                        if(resp.getPollStatus().getDeviceConfig() != null) {
-                            return resp.getPollStatus().getDeviceConfig().content;
-                        }
-                        return null;
-                })
+                .thenApply(resp -> resp.getPollStatus().getDeviceConfig())
                 .whenComplete((config, throwable) -> {
                     if (throwable != null) {
                         LOG.error("Error while getting device config for IpAddress {} at location {}", ipAddress, location, throwable);
                     }
                 });
     }
 
-    private CompletableFuture<PollerResponse> pollDeviceConfig(String ipAddress, String location, String configType) {
+    private CompletableFuture<PollerResponse> pollDeviceConfig(String ipAddress, String location, String configType) throws IOException {
         if (Strings.isNullOrEmpty(configType)) {
             configType = ConfigType.Default;
         }
         String serviceName = DEVICE_CONFIG_SERVICE_NAME_PREFIX + configType;
+        final String configTypeName = configType;
         MonitoredService service = sessionUtils.withReadOnlyTransaction(() -> {
-
             OnmsIpInterface ipInterface = ipInterfaceDao.findByIpAddressAndLocation(ipAddress, location);
             if (ipInterface == null) {
                 return null;
             }
             OnmsNode node = ipInterface.getNode();
-
+            List<OnmsMonitoredService> services =
+                    monitoredServiceDao.findSimilarServicesOnInterface(node.getId(), ipInterface.getIpAddress(), DEVICE_CONFIG_SERVICE_NAME_PREFIX);
+            if (!services.isEmpty()) {
+                Optional<OnmsMonitoredService> optional =
+                        services.stream().filter(onmsMonitoredService -> onmsMonitoredService.getServiceName().contains(configTypeName)).findFirst();
+                if (optional.isPresent()) {
+                    return new SimpleMonitoredService(ipInterface.getIpAddress(), node.getId(), node.getLabel(), optional.get().getServiceName(), location);
+                }
+            }
             return new SimpleMonitoredService(ipInterface.getIpAddress(), node.getId(), node.getLabel(), serviceName, location);
         });
 
         if (service == null) {
             throw new IllegalArgumentException("No interface found with ipAddress " + ipAddress + " at location " + location);
         }
+        List<Parameter> parameters = getServiceParamsFromPollerConfig(service);
+        Map<String, Object> serviceAttributes = convertParamsToAttributes(parameters);
```

Comment:
Should this do metadata interpolation instead?

Reply:
`PollerRequestBuilderImpl` does interpolation in execution.
 This only needs to retrieve service configuration from Poller configuration.